alphagov/paas-billing
# [#158665945] Use plans `unique_id` to identify the pricing_plans

What
----

This PR changes how we map the service plans for billing purposes with the events. 

Currently we were using the Cloud Foundry generated `plan_guid` for this purpose. This had the limitation that different configurations of the service plans would be needed on a per environment basis.

Since we did the work to collect the historic data in https://github.com/alphagov/paas-billing/pull/36 we have the broker reported plan unique IDs available for this purpose.

This PR would leverage the table `service_plans` to map the events to the corresponding plans using the plan unique ID reported by the broker. It also updates the embedded `config.json` to use the `unique_id` instead of the specific CF plan id.

Notes about the implementation: 

 - Old Pricing Calculator URLs would break:
    - We change the `config.json` IDs, which means that the `/pricing_plans` would report a different GUID for the plans.
   - The forecast events would use the `unique_id`. This makes sense as the `/pricing_plans` where `paas-admin` calculator gets the values would be using `uinque_ids`.
 
 - We add some additional tests to cover this case
 - We add logic to map any service not stored in the historic tables `services` and `service_plans` to a "Unknown Plan" with cost 0.
 - We added a SQL script to add a missing service plan for `redis tiny (compose)`.


How to review
-----

- Code review
- Deploy https://github.com/alphagov/paas-cf/pull/1502
- Check you have bills for services in your dev env, admin is a good org to test with.

Optionally testing in prod data (requires access to prod): 
 - We did compare all the costs calculated with this version for all the orgs in prod.
 - All the notes are in the private HackMD of the story. The scripts are documented in a comment below.: 

Before merge
----------------

It is important to run the script: `scripts/create_adhoc_redis_service.sql` in prod before deploying in prod:

```
cf login -a https://api.cloud.service.gov.uk --sso
cf target -o admin -s billing
cf conduit billing-db -- psql 
```
and paste the script.

Who can review
-----

Not @LeePorte or @keymon
